engine/crates/engine-v2/src/operation/bind/validation.rs

```diff
@@ -0,0 +1,133 @@
+use config::latest::OperationLimits;
+use engine::Positioned;
+
+use crate::operation::parse::ParsedOperation;
+
+use super::{BindError, BindResult};
+
+pub(super) fn validate_parsed_operation(operation: &ParsedOperation, limits: &OperationLimits) -> BindResult<()> {
+    Visitor {
+        operation,
+        current_fragments_stack: Vec::new(),
+        root_fields: 0,
+        max_root_fields: limits.root_fields.map(Into::into).unwrap_or(usize::MAX),
+        current_depth: 0,
+        max_depth: limits.depth.map(Into::into).unwrap_or(usize::MAX),
+        aliases_count: 0,
+        max_aliases_count: limits.aliases.map(Into::into).unwrap_or(usize::MAX),
+        complexity: 0,
+        max_complexity: limits.complexity.map(Into::into).unwrap_or(usize::MAX),
+    }
+    .visit_selection_set(true, &operation.definition.selection_set)
+}
+
+struct Visitor<'p> {
+    operation: &'p ParsedOperation,
+    current_fragments_stack: Vec<&'p str>,
+    root_fields: usize,
+    max_root_fields: usize,
+    current_depth: usize,
+    max_depth: usize,
+    aliases_count: usize,
+    max_aliases_count: usize,
+    complexity: usize,
+    max_complexity: usize,
+}
+
+impl<'p> Visitor<'p> {
+    fn visit_selection_set(
+        &mut self,
+        root: bool,
+        selection_set: &'p Positioned<engine_parser::types::SelectionSet>,
+    ) -> BindResult<()> {
+        for item in &selection_set.items {
+            match &item.node {
+                engine_parser::types::Selection::Field(field) => {
+                    self.root_fields += root as usize;
```

Comment:
Isn't root fields only the top-most selection set?

Reply:
(if so, this can be wrapped in an `if current_depth == 0`)

Reply:
you're right